File: channel-core/test/signature_decode_strictness.test.ts
import { describe, expect, test, vi } from "vitest";
import * as ed from "@noble/ed25519";
import { sha512 } from "@noble/hashes/sha2.js";
import { computeDIDKey } from "../src/identity/did.js";
import {
  canonicalJSON,
  signMessage,
  verifyMessage,
  verifySignedPayload,
  type MessageEnvelope,
} from "../src/identity/signing.js";

// docs/vectors/README.md requires that identity/signing.ts throw on malformed
// signature base64 before calling Ed25519 verification, so that rejection does
// not depend on the verifier's treatment of an empty signature. ed.verify
// returns false rather than throwing on garbage, so "failed" alone cannot
// distinguish a decoder rejection from a verifier rejection. These tests
// delegate to the real verifier while counting whether it was reached at all.
const verifyCalls = vi.hoisted(() => vi.fn());
vi.mock("@noble/ed25519", async (importOriginal) => {
  const actual = await importOriginal<typeof import("@noble/ed25519")>();
  return {
    ...actual,
    verify: (...args: Parameters<typeof actual.verify>) => {
      verifyCalls();
      return actual.verify(...args);
    },
  };
});

ed.etc.sha512Sync = (...m) => sha512(ed.etc.concatBytes(...m));

async function signedEnvelope(): Promise<{ envelope: MessageEnvelope; signature: string }> {
  const seed = new Uint8Array(32).fill(7);
  const did = computeDIDKey(await ed.getPublicKeyAsync(seed));
  const envelope: MessageEnvelope = {
    from: "mycompany/researcher",
    from_did: did,
    to: "otherco/monitor",
    to_did: "",
    type: "mail",
    subject: "task complete",
    body: "results attached",
    timestamp: "2026-02-21T15:30:00Z",
  };
  const signature = await signMessage(seed, envelope);
  return { envelope: { ...envelope, signing_key_id: did }, signature };
}

describe("signature decoding rejects malformed base64 before verification", () => {
  // atob accepts padding, so without the strict alphabet check a signature Go
  // rejects would decode to the correct 64 bytes and verify here. Each test
  // ends by verifying the unmodified signature, which proves the fixture is a
  // signature a lenient decoder would accept and that the decode branch is
  // reached at all.
  test("padded signatures never reach the verifier", async () => {
    const { envelope, signature } = await signedEnvelope();

    verifyCalls.mockClear();
    expect(await verifyMessage({ ...envelope, signature: `${signature}==` })).toBe("failed");
    expect(verifyCalls).not.toHaveBeenCalled();

    expect(await verifyMessage({ ...envelope, signature })).toBe("verified");
    expect(verifyCalls).toHaveBeenCalledTimes(1);
  });

  test("padded signatures never reach the verifier via verifySignedPayload", async () => {
    const { envelope, signature } = await signedEnvelope();
    const payload = canonicalJSON(envelope);
    const did = envelope.from_did;

    verifyCalls.mockClear();
    expect(await verifySignedPayload(payload, `${signature}==`, did, did)).toBe("failed");
    expect(verifyCalls).not.toHaveBeenCalled();

    expect(await verifySignedPayload(payload, signature, did, did)).toBe("verified");
    expect(verifyCalls).toHaveBeenCalledTimes(1);
  });

  // Go's RawStdEncoding skips only \r and \n; atob strips all whitespace. A
  // space-bearing signature is therefore accepted by atob and rejected by Go,
  // the same acceptance divergence as padding.
  test("signatures carrying interior whitespace never reach the verifier", async () => {
    const { envelope, signature } = await signedEnvelope();
    const spaced = `${signature.slice(0, 4)} ${signature.slice(4)}`;

    verifyCalls.mockClear();
    expect(await verifyMessage({ ...envelope, signature: spaced })).toBe("failed");
    expect(verifyCalls).not.toHaveBeenCalled();

    expect(await verifyMessage({ ...envelope, signature })).toBe("verified");
    expect(verifyCalls).toHaveBeenCalledTimes(1);
  });

  // Pins the decoder choice at signing.ts b64Decode: Buffer.from(value,
  // "base64") silently drops characters outside the alphabet and would hand
  // the verifier seven bytes of salvage here.
  test("non-alphabet signatures never reach the verifier", async () => {
    const { envelope } = await signedEnvelope();

    verifyCalls.mockClear();
    expect(await verifyMessage({ ...envelope, signature: "not-base64!" })).toBe("failed");
    expect(verifyCalls).not.toHaveBeenCalled();
  });
});

File: cli/go/awid/signature_decode_strictness_test.go
package awid

import (
	"crypto/ed25519"
	"encoding/base64"
	"errors"
	"testing"
)

// VerifyMessage and VerifySignedPayload decode the signature with
// RawStdEncoding, which rejects the padded input StdEncoding accepts.
// ed25519.Verify returns false rather than erroring on garbage, so a status of
// Failed on its own cannot distinguish "the decoder rejected the signature"
// from "the signature decoded and did not verify" — both are Failed. These
// tests discriminate the two by the returned error: only the decode path
// returns a base64.CorruptInputError, and a verifier rejection returns Failed
// with a nil error (TestSignatureVerifierRejectionCarriesNoError pins that
// half). Each padded-signature test also asserts the unpadded signature
// verifies, so the fixture is known to be a signature a lenient decoder would
// accept and the decode branch is known to be reached.

func signedTestEnvelope(t *testing.T) (*MessageEnvelope, string, string) {
	t.Helper()

	pub, priv, err := ed25519.GenerateKey(nil)
	if err != nil {
		t.Fatal(err)
	}
	did := ComputeDIDKey(pub)

	env := &MessageEnvelope{
		From:      "mycompany/researcher",
		FromDID:   did,
		To:        "otherco/monitor",
		ToDID:     "did:key:z6Mkf5rGMoatrSj1f4CyvuHBeXJELe9RPdzo2PKGNCKVtZxP",
		Type:      "mail",
		Subject:   "task complete",
		Body:      "results attached",
		Timestamp: "2026-02-21T15:30:00Z",
	}

	sig, err := SignMessage(priv, env)
	if err != nil {
		t.Fatalf("SignMessage: %v", err)
	}
	env.SigningKeyID = did
	return env, sig, did
}

func requireCorruptInput(t *testing.T, err error) {
	t.Helper()

	if err == nil {
		t.Fatal("err=nil, want a base64 decode error: the signature reached the verifier instead of being rejected at decode")
	}
	var corrupt base64.CorruptInputError
	if !errors.As(err, &corrupt) {
		t.Fatalf("err=%v, want a base64.CorruptInputError: rejection did not come from the signature decoder", err)
	}
}

func TestVerifyMessageRejectsPaddedSignatureAtDecode(t *testing.T) {
	t.Parallel()

	env, sig, _ := signedTestEnvelope(t)

	env.Signature = sig + "=="
	status, err := VerifyMessage(env)
	if status != Failed {
		t.Fatalf("padded signature: status=%q, want %q", status, Failed)
	}
	requireCorruptInput(t, err)

	env.Signature = sig
	if status, err := VerifyMessage(env); status != Verified || err != nil {
		t.Fatalf("unpadded signature: status=%q err=%v, want %q with no error", status, err, Verified)
	}
}

func TestVerifySignedPayloadRejectsPaddedSignatureAtDecode(t *testing.T) {
	t.Parallel()

	env, sig, did := signedTestEnvelope(t)
	payload := CanonicalJSON(env)

	status, err := VerifySignedPayload(payload, sig+"==", did, did)
	if status != Failed {
		t.Fatalf("padded signature: status=%q, want %q", status, Failed)
	}
	requireCorruptInput(t, err)

	if status, err := VerifySignedPayload(payload, sig, did, did); status != Verified || err != nil {
		t.Fatalf("unpadded signature: status=%q err=%v, want %q with no error", status, err, Verified)
	}
}

// A signature that decodes cleanly but does not verify must return Failed with
// a nil error. Without this, the decode tests above could not treat a non-nil
// error as evidence that the verifier was never reached.
func TestSignatureVerifierRejectionCarriesNoError(t *testing.T) {
	t.Parallel()

	env, sig, did := signedTestEnvelope(t)
	payload := CanonicalJSON(env)
	env.Signature = sig
	env.Body = "tampered body"

	status, err := VerifyMessage(env)
	if status != Failed {
		t.Fatalf("VerifyMessage status=%q, want %q", status, Failed)
	}
	if err != nil {
		t.Fatalf("VerifyMessage err=%v, want nil for a crypto rejection", err)
	}

	status, err = VerifySignedPayload(payload+" ", sig, did, did)
	if status != Failed {
		t.Fatalf("VerifySignedPayload status=%q, want %q", status, Failed)
	}
	if err != nil {
		t.Fatalf("VerifySignedPayload err=%v, want nil for a crypto rejection", err)
	}
}
